### PR TITLE
 [DS-3306] bin/dspace create-administrator needs a --help option #6661 #8367

### LINKED ISSUE #6661: [DS-3306] bin/dspace create-administrator needs a --help option
<i>Imported from JIRA <a href="https://jira.lyrasis.org/browse/DS-3306">[DS-3306]</a> created by mwood</i>
<p>create-administrator doesn't respond well to any of the usual methods of asking how to use it.  It should be made to understand -h, --help, and -?</p>

<p>While the code is open, also it should not require -c if webui.supported.locales is empty.</p>

<p>Oh, one more thing:  we should just prompt for the password if it isn't specified but there are enough other options to create the user.</p>

### PR DESCRIPTION
**Pull request by ammanabrolua**
## References
Fixes #6661

## Description
Added create-administrator script changes to add the -h flag

## Checklist
1. -h flag usage
2. prompt for password if not already not provided with -e, -f, -l and -c options
3. -c required when webui.supported.locales is not null
4. -c not required when webui.supported.locales is null

### CODE REVIEW

File: dspace-api/src/main/java/org/dspace/administer/CreateAdministrator.java [hunk 1]
```diff
@@ -64,24 +66,115 @@ public static void main(String[] argv)
...
+
+                        language = console.readLine();
+
+                        if (language != null) {
+                            language = language.trim();
+                            language = I18nUtil.getSupportedLocale(new Locale(language)).getLanguage();
+                        }
+                    }
+
+                    while (!dataOK) {
```
Comment (mwoodiupui) [inline dspace-api/src/main/java/org/dspace/administer/CreateAdministrator.java:138]: This loop only allows the user to correct the password, which cannot be seen.  The original will ask for all of the user properties again.  Neither is what I would expect.

Suggestion:  factor out the password-prompting code into a new private method.  Loop there, not returning until both passwords match.  Return one of the character arrays after clearing the other one.  Call from `negotiateAdministratorDetails`.

Suggestion:  `negotiateAdministratorDetails` should display the previously-collected values (if any) when looping through its prompts.  An empty response to a prompt should keep the existing value.

File: dspace-api/src/main/java/org/dspace/administer/CreateAdministrator.java [hunk 2]
```diff
@@ -64,24 +66,115 @@ public static void main(String[] argv)
...
+                    " DSpace site. Unless all the required parameters are passed it will" +
+                    " prompt for an e-mail address, last name, first name and password from" +
+                    " standard input.. An administrator group is then created and the data passed" +
+                    "  in used to create an e-person in that group.\n\n";
+                String footer = "\n";
+                HelpFormatter formatter = new HelpFormatter();
+                formatter.printHelp("dspace create-administrator", header, options, footer, true);
+                return;
+            } else {
+
```
[the remark this answers is not in the document]
Comment (ammanabrolua) [reply dspace-api/src/main/java/org/dspace/administer/CreateAdministrator.java:116]: > Why duplicate `negotiateAdministratorDetails`?

Prompt user for password if the -p is missing.

[the remark this answers is not in the document]
Comment (ammanabrolua) [reply dspace-api/src/main/java/org/dspace/administer/CreateAdministrator.java:116]: > Why duplicate `negotiateAdministratorDetails`?

If all the options are not satisfied in the above conditionals then call ```negotiateAdministratorDetails```.

File: dspace-api/src/main/java/org/dspace/administer/CreateAdministrator.java (hunk 1 above)
Comment (ammanabrolua) [reply dspace-api/src/main/java/org/dspace/administer/CreateAdministrator.java:138]: I am working on the changes. Will take sometime.

Comment (ammanabrolua) [reply dspace-api/src/main/java/org/dspace/administer/CreateAdministrator.java]: I changed the code

File: dspace-api/src/main/java/org/dspace/administer/CreateAdministrator.java [hunk 3]
```diff
@@ -114,91 +144,124 @@ protected void negotiateAdministratorDetails()
...
         String lastName = null;
-        char[] password1 = null;
-        char[] password2 = null;
         String language = I18nUtil.getDefaultLocale().getLanguage();
+        ConfigurationService cfg = DSpaceServicesFactory.getInstance().getConfigurationService();
+        boolean flag = false;
+        char[] password = null;
+
+
+        if (!line.hasOption('p') && line.hasOption("e")  && line.hasOption("f") && line.hasOption("l")
```
Comment (mwoodiupui) [inline dspace-api/src/main/java/org/dspace/administer/CreateAdministrator.java:153]: The logic here is difficult to understand.  I think that `flag` means "we don't have a password yet" and `dataOK` means "we have all required information other than the password."  It might be clearer to test for these separately.

There are four cases here:  (1) user gave all required values except the password; (2) user gave a password and NONE of the other required values; (3) user gave all required values; (4) [drop off the end of the `if`] some other combination.  I think that we need only three cases, and by separating the password question we need only two.

First set `flag = line.hasOption('p');`.  Then set `dataOK` true if all other required properties are set.  Then `!flag` means "we need to prompt for a password" and `!dataOK` means "we need to prompt for the other required properties."  Is that how you meant it to work?

File: dspace-api/src/main/java/org/dspace/administer/CreateAdministrator.java [hunk 4]
```diff
@@ -114,91 +144,124 @@ protected void negotiateAdministratorDetails()
...
-                System.out.println("Please provide an email address.");
-                continue;
-            }
-
-            System.out.print("First name: ");
-            System.out.flush();
-
-            firstName = console.readLine();
+            if (flag) {
+                password = getPassword(console);
```
Comment (mwoodiupui) [inline dspace-api/src/main/java/org/dspace/administer/CreateAdministrator.java:170]: You might want to pull the password prompt up out of the loop.  The user has no way to check whether the password he typed is the one that he meant, so there's no way to correct it.  The password has a different check:  was it entered the same way twice?

File: dspace-api/src/main/java/org/dspace/administer/CreateAdministrator.java [hunk 5]
```diff
@@ -114,91 +144,124 @@ protected void negotiateAdministratorDetails()
...
+                if (password != null) {
+                    dataOK = true;
+                }
+            } else if (!flag) {
```
Comment (mwoodiupui) [inline dspace-api/src/main/java/org/dspace/administer/CreateAdministrator.java:174]: The `if (!flag)` is redundant.  If we got here, `flag` is false.

File: dspace-api/src/main/java/org/dspace/administer/CreateAdministrator.java [hunk 6]
```diff
@@ -114,91 +144,124 @@ protected void negotiateAdministratorDetails()
...
+                        language = I18nUtil.getSupportedLocale(new Locale(language)).getLanguage();
+                    }
+                }
 
-            //TODO real password validation
-            if (password1.length > 1 && Arrays.equals(password1, password2)) {
-                // password OK
-                System.out.print("Is the above data correct? (y or n): ");
-                System.out.flush();
+                password = getPassword(console);
```
Comment (mwoodiupui) [inline dspace-api/src/main/java/org/dspace/administer/CreateAdministrator.java:218]: This becomes redundant if the password prompt is moved out of the loop.

File: dspace-api/src/main/java/org/dspace/administer/CreateAdministrator.java (hunk 3 above)
Comment (ammanabrolua) [reply dspace-api/src/main/java/org/dspace/administer/CreateAdministrator.java:153]: Prompt for password when not provided but all the other fields are provided

File: dspace-api/src/main/java/org/dspace/administer/CreateAdministrator.java (hunk 4 above)
Comment (ammanabrolua) [reply dspace-api/src/main/java/org/dspace/administer/CreateAdministrator.java:170]: will look into this

File: dspace-api/src/main/java/org/dspace/administer/CreateAdministrator.java (hunk 5 above)
Comment (ammanabrolua) [reply dspace-api/src/main/java/org/dspace/administer/CreateAdministrator.java:174]: will look into this

File: dspace-api/src/main/java/org/dspace/administer/CreateAdministrator.java (hunk 6 above)
Comment (ammanabrolua) [reply dspace-api/src/main/java/org/dspace/administer/CreateAdministrator.java:218]: will have to look into this

**Review by mwoodiupui — CHANGES_REQUESTED** [review]
I found the logic hard to follow, and I think the reason for this is that the password must be treated separately from the other required properties because it is prompted and accepted in a different manner, but this is then mixed in with the other properties.

**Comment by ammanabrolua** [comment]
> I found the logic hard to follow, and I think the reason for this is that the password must be treated separately from the other required properties because it is prompted and accepted in a different manner, but this is then mixed in with the other properties.

Hi @mwoodiupui . Please have a look at the modified code.

File: dspace-api/src/main/java/org/dspace/administer/CreateAdministrator.java [hunk 7]
```diff
@@ -103,20 +126,20 @@ protected CreateAdministrator()
...
-        throws Exception {
+    protected void negotiateAdministratorDetails(CommandLine line)
+            throws Exception {
         Console console = System.console();
 
         System.out.println("Creating an initial administrator account");
 
-        boolean dataOK = false;
-
         String email = null;
```
Comment (mwoodiupui) [inline dspace-api/src/main/java/org/dspace/administer/CreateAdministrator.java:135]:
```
String email = line.getOptionValue('e');
String firstName = line.getOptionValue('f');
String lastName = line.getOptionValue('l');
```
I tested this patch giving all three options plus `-p` and got an EPerson with no email address.

File: dspace-api/src/main/java/org/dspace/administer/CreateAdministrator.java [hunk 8]
```diff
@@ -103,20 +126,20 @@ protected CreateAdministrator()
...
         String firstName = null;
         String lastName = null;
-        char[] password1 = null;
-        char[] password2 = null;
         String language = I18nUtil.getDefaultLocale().getLanguage();
+        ConfigurationService cfg = DSpaceServicesFactory.getInstance().getConfigurationService();
+        boolean flag = line.hasOption('p');
+        char[] password = null;
+        boolean dataOK = !(!line.hasOption('f') || !line.hasOption('e') || !line.hasOption('l'));
```
Comment (mwoodiupui) [inline dspace-api/src/main/java/org/dspace/administer/CreateAdministrator.java:142]: By deMorgan's Rule this reduces to `boolean dataOK = line.hasOption('f') && line.hasOption('e') && line.hasOption('l');` which is a bit easier to understand.

File: dspace-api/src/main/java/org/dspace/administer/CreateAdministrator.java (hunk 7 above)
Comment (mwoodiupui) [reply dspace-api/src/main/java/org/dspace/administer/CreateAdministrator.java:135]: I tried my suggested change and got the result I expected.

Comment (ammanabrolua) [reply dspace-api/src/main/java/org/dspace/administer/CreateAdministrator.java:135]: I did not understand this point. Even after demorgan's changes this will not work?

Comment (ammanabrolua) [reply dspace-api/src/main/java/org/dspace/administer/CreateAdministrator.java:135]: @mwoodiupui . Please help me understand.

Comment (ammanabrolua) [reply dspace-api/src/main/java/org/dspace/administer/CreateAdministrator.java]: Dear @mwoodiupui ,

I changed:

```String email = null;
String firstName = null;
String lastName = null;```

to:

```
String email = line.getOptionValue('e');
String firstName = line.getOptionValue('f');
String lastName = line.getOptionValue('l');
```
Is this what you meant?

File: dspace-api/src/main/java/org/dspace/administer/CreateAdministrator.java (hunk 8 above)
Comment (ammanabrolua) [reply dspace-api/src/main/java/org/dspace/administer/CreateAdministrator.java]: @mwoodiupui . I resolved this

File: dspace-api/src/main/java/org/dspace/administer/CreateAdministrator.java (hunk 7 above)
Comment (mwoodiupui) [reply dspace-api/src/main/java/org/dspace/administer/CreateAdministrator.java]: Yes, that is what I meant.  Thank you.

**Review by mwoodiupui — APPROVED** [review]

**Review by tdonohue — APPROVED** [review]
👍 Thanks @ammanabrolua ! This looks good to me too.
